### DISCUSSION TITLE
`if(!_isApprovedOrOwner(msg.sender, tokenId))`

### DISCUSSION THREAD

**Discussion by Not-Sarthak** [Cyfrin/foundry-full-course-cu#2267] [Q&A] [answered]
In SVG NFTs section, Patrick has used `if(!_isApprovedOrOwner(msg.sender, tokenId))` to check whether the NFT being flipped is done by the Owner or not. but in the example Github Repo: https://github.com/Cyfrin/foundry-nft-cu/blob/main/test/MoodNftTest.sol it uses ` if (getApproved(tokenId) != msg.sender && ownerOf(tokenId) != msg.sender)` for the same thing. 

What is the difference between the both?

Here's the function:
```
function flipMood(uint256 tokenId) public {
		    // if(!_isApprovedOrOwner(msg.sender, tokenId))
        if (getApproved(tokenId) != msg.sender && ownerOf(tokenId) != msg.sender) {
            revert MoodNft__CantFlipMoodIfNotOwner();
        }

        if (s_tokenIdToState[tokenId] == NFTState.HAPPY) {
            s_tokenIdToState[tokenId] = NFTState.SAD;
        } else {
            s_tokenIdToState[tokenId] = NFTState.HAPPY;
        }
    }```

**Comment by EngrPips** [comment, marked as answer]
Hello @Not-Sarthak, `_isApprovedOrOwner` checks if the `msg.sender` is either the owner of the `NFT` or an approved user by the owner while `getApproved(tokenId) != msg.sender && ownerOf(tokenId) != msg.sender` split the functionality so that `getApproved` check if the `msg.sender` is the owner of the `NFT` and `ownerOf`  checks if the `msg.sender` is the owner of the `NFT`.

**Reply by EngrPips** [reply]
In programming, `separation of concern` seems to be one of those things that are considered as good programming habit. Ideally a function should do only one thing and do it well.

**Comment by Not-Sarthak** [comment]
Thanks a lot! that makes sense.

**Reply by EngrPips** [reply]
You are welcome buddy, Keep crushing it.

**Comment by BowtiedHarpyEagle** [comment]
I also think _isApprovedOrOwner() is no longer part of the ERC721 openzeppelin library. I had a red error mark on that part, came here to ask a question, found this.

**Reply by EngrPips** [reply]
Interesting.
